### ISSUE TITLE
Unable to set checkbox.cascade with three_state true

### ISSUE THREAD

**Issue by abeverley** [vakata/jstree#1955]
I am unable to configure $.jstree.defaults.checkbox.cascade when $.jstree.defaults.checkbox.three_state is true.

From what I can see in the code, this is because the default value of $.jstree.defaults.checkbox.cascade (set to 'up+down+undetermined' if three_state is true) is set after the options have been read, regardless of whether cascade already has a value:

```
    $.jstree.plugins.checkbox = function (options, parent) {
        this.bind = function () {
            ...
            if(this.settings.checkbox.three_state) {
                this.settings.checkbox.cascade = '';
            }
```

I have worked around it for now by doing this:

    $('#jstree').jstree(true).settings.checkbox.cascade = 'undetermined';

Thanks,

Andy

**Comment by vakata** [comment]
This is by design. ```three_state``` is by far the more used option, and was available first. So to use ```cascade``` you should set ```three_state``` to ```false```. It is mentioned in the docs: https://www.jstree.com/api/#/?q=cascade&f=$.jstree.defaults.checkbox.cascade

**Comment by abeverley** [comment]
Thanks for the quick reply. I still can't get the configuration I want without the hack though:

What if I want to enable "undetermined state" checkboxes, but only have an "undetermined" cascade happen?

What I want, is for collapsed nodes that have a value checked underneath then, to automatically show the undetermined state, but without cascading all values when a node is selected. This is only possible with a "checkbox.cascade" configuration value of "undetermined" and a "checkbox.three_state" option of "true". It's not possible to set configuration options to those values.

Why not allow checkbox.cascade to be overridden, and only set the automatic "up+down+undetermined" value if it hasn't been defined in the options?

**Comment by vakata** [comment]
As I mentioned - ```three_state``` is just a shortcut (as you can see in the code) which sets ```cascade``` to ```up+down+undetermined```. All available options are in the cascade option. 

My understanding is that you do not want to propagate clicks up and down, but have the visual indication of undetermined nodes as if clicks would propagate. Unfortunately this is not something you can achieve with the current set of options - undetermined nodes are used only when there is propagation.
What you can do is listen for changed events and after a small timeout - add the undetermined class to all nodes you think should be in that state. 

I can not help you further because I do not understand your requirements - what I understood makes no sense - if a node should be undetermined, when nodes below it are checked, but clicking that has no effect on the nodes below it, does it become checked? How about deselection - when deselecting that same node - does it become unchecked or reverses back to undetermined? What happens when all its children are checked - does the parent become checked or unchecked? For me undetermined nodes make sense only if state is autocalculated and propagated.

**Comment by abeverley** [comment]
Thanks for the comprehensive reply @vakata :

> As I mentioned - three_state is just a shortcut (as you can see in the code) which sets cascade to up+down+undetermined. All available options are in the cascade option.

Thanks - that was my misunderstanding. When the docs said "this setting is automatically set to 'up+down+undetermined'" I thought that means it is set as the default (which could be overridden) rather than a hard value.

> My understanding is that you do not want to propagate clicks up and down, but have the visual indication of undetermined nodes as if clicks would propagate.

Correct.

> Unfortunately this is not something you can achieve with the current set of options - undetermined nodes are used only when there is propagation.

Thanks for the clarification - I didn't realise this was an unsupported configuration.

> What you can do is listen for changed events and after a small timeout - add the undetermined class to all nodes you think should be in that state.

Thanks for the advice. I've managed to achieve what I want by short-cutting the configuration options as per my initial comment. It would be nice if this combination of options was supported, as it works well enough for my needs.

> I can not help you further because I do not understand your requirements - what I understood makes no sense - if a node should be undetermined, when nodes below it are checked, but clicking that has no effect on the nodes below it, does it become checked?

My requirement is to inform the user visually that there are *some* selected nodes below a closed node. Otherwise, a user sees a closed node and assumes nothing below it is selected, when actually something might be selected. However, I do not want cascading to happen 

> How about deselection - when deselecting that same node - does it become unchecked or reverses back to undetermined?

I only actually want a user to be able to select end nodes, so I'm currently happy with clicking a parent node to just open and close the branch below it.

> What happens when all its children are checked - does the parent become checked or unchecked?

Ideally all children checked = parent is checked, but this is the one part of my workaround that doesn't work, as the parent node stays undetermined. However, this behaviour is still preferable to the alternatives.

> For me undetermined nodes make sense only if state is autocalculated and propagated. If you need a way to indicate to the user that something below an unchecked node is checked (only thing I could think of), in my opinion this is not the proper way to do it.

This is exactly it, otherwise users get confused because in a large tree they don't realise that something below a closed unchecked parent is checked.

In short, it would be good to allow checkbox.cascade to be overridden even when checkbox.three_state is true. This would still be backwards-compatible. Specifically:

```
-    if(this.settings.checkbox.three_state) {
+    if(this.settings.checkbox.three_state && !this.settings.checkbox.cascade) {
        this.settings.checkbox.cascade = '';
    }
```

Thanks, Andy
